pkg/rollout/rollout.go

```diff
@@ -97,35 +103,64 @@ func (r *Rollout) UpdateService(svc *run.Service) (*run.Service, error) {
 		return nil, nil
 	}
 	r.log.Debugf("%q is the stable revision", stable)
+	r.log = r.log.WithField("stable", stable)
 
 	candidate := DetectCandidateRevisionName(svc, stable)
 	if candidate == "" {
 		r.log.Info("could not determine candidate revision")
 		return nil, nil
 	}
-	r.log.Debugf("%q is the candidate revision", candidate)
+	r.log.Debugf("%q is the candidate revision", stable)
+	r.log = r.log.WithField("candidate", candidate)
+
+	// A new candidate does not have metrics yet, so it can't be diagnosed.
+	if isNewCandidate(svc, candidate) {
+		r.log.Debug("new candidate found")
+		svc = r.RollForward(svc, stable, candidate)
+		svc = r.updateAnnotations(svc, stable, candidate)
+		return r.replaceService(svc)
+	}
 
-	svc = r.SplitTraffic(svc, stable, candidate)
-	svc = r.updateAnnotations(svc, stable, candidate)
-	svc, err := r.runClient.ReplaceService(r.project, r.serviceName, svc)
+	diagnosis, err := r.diagnoseCandidate(candidate, r.strategy.Metrics)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not update service %q", r.serviceName)
+		r.log.Error("could not diagnose candidate's health")
+		return nil, errors.Wrapf(err, "failed to diagnose health for candidate %q", candidate)
 	}
-	r.log.Debug("service succesfully updated")
 
-	return svc, nil
+	switch diagnosis.OverallResult {
+	case health.Unknown:
+		r.log.Error("unknown candidate's health")
+		return nil, errors.Errorf("got unknown health for candidate %q", candidate)
+	case health.Inconclusive:
+		r.log.Debug("no enough requests to determine health, skipping rollout/rollback for now")
+		return nil, nil
+	case health.Healthy:
+		r.log.Debug("candidate is healthy, will increase traffic to candidate")
+		svc = r.RollForward(svc, stable, candidate)
+		break
+	case health.Unhealthy:
+		r.log.Infof("candidate did not meet health criteria, will roll back to %q", stable)
```

Comment:
nit: Reading logs is difficult. They are more learnable when logs are succinct and different messages for variants of the same status use the same grammar.

```
candidate is healthy, will increase traffic to candidate
candidate is not healthy, will roll back to %q
candidate is missing health criteria, skipping rollout/rollback
```

In skimming logs, the reader knows they are searching for "candidate is..." and from their will be able to identify the necessary state.

Reply:
I really like this. I'll keep it in mind. 

In this case, I'm not particularly convinced by `candidate is missing health criteria, skipping rollout/rollback` since it doesn't capture the idea of enough requests. Let me see if I can think of a different phrase.